File: ios/Sources/VaultSync/Presentation/ViewModels/ConflictsViewModel.swift
import SwiftUI
import Combine

@MainActor
public final class ConflictsViewModel: ObservableObject {
    @Published public var conflicts: [Conflict] = []

    private let syncRepository: any SyncRepository
    private let resolveConflictUseCase: ResolveConflictUseCase

    public init(
        syncRepository: any SyncRepository,
        resolveConflictUseCase: ResolveConflictUseCase
    ) {
        self.syncRepository = syncRepository
        self.resolveConflictUseCase = resolveConflictUseCase

        Task {
            await observeConflicts()
        }
    }

    public func observeConflicts() async {
        for await list in syncRepository.observeConflicts() {
            self.conflicts = list
        }
    }

    public func resolve(conflictId: String, resolution: ConflictResolution) {
        Task {
            try? await resolveConflictUseCase.execute(conflictId: conflictId, resolution: resolution)
        }
    }
}

File: android/app/src/main/java/com/vaultsync/ui/viewmodel/ConflictsViewModel.kt
package com.vaultsync.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vaultsync.domain.model.Conflict
import com.vaultsync.domain.model.ConflictResolution
import com.vaultsync.domain.repository.SyncRepository
import com.vaultsync.domain.usecase.ResolveConflictUseCase
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class ConflictsViewModel(
    syncRepository: SyncRepository,
    private val resolveConflictUseCase: ResolveConflictUseCase
) : ViewModel() {

    val conflicts: StateFlow<List<Conflict>> = syncRepository.observeConflicts()
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = emptyList()
        )

    fun resolve(conflictId: String, resolution: ConflictResolution) {
        viewModelScope.launch {
            resolveConflictUseCase(conflictId, resolution)
        }
    }
}
